File: engine/sqlite_index.js
// SQLite FTS5 index for the Local Context Engine (JS port).
// Uses Node's built-in `node:sqlite` (available in Node 22.5+; tested on Node 24).

import { DatabaseSync } from "node:sqlite";
import { mkdirSync } from "node:fs";
import { dirname } from "node:path";
import { Chunk, SourceKind } from "./models.js";
import { tokenize } from "./text_utils.js";

const SCHEMA = `
CREATE TABLE IF NOT EXISTS chunks (
    rowid INTEGER PRIMARY KEY AUTOINCREMENT,
    id TEXT UNIQUE NOT NULL,
    source TEXT NOT NULL,
    kind TEXT NOT NULL,
    content TEXT NOT NULL,
    metadata TEXT NOT NULL,
    timestamp REAL,
    priority REAL,
    tokens INTEGER NOT NULL
);

CREATE VIRTUAL TABLE IF NOT EXISTS chunks_fts USING fts5(
    content,
    content='chunks',
    content_rowid='rowid',
    tokenize = 'unicode61'
);

CREATE TRIGGER IF NOT EXISTS chunks_ai AFTER INSERT ON chunks BEGIN
    INSERT INTO chunks_fts(rowid, content) VALUES (new.rowid, new.content);
END;

CREATE TRIGGER IF NOT EXISTS chunks_ad AFTER DELETE ON chunks BEGIN
    INSERT INTO chunks_fts(chunks_fts, rowid, content)
    VALUES ('delete', old.rowid, old.content);
END;

CREATE TRIGGER IF NOT EXISTS chunks_au AFTER UPDATE ON chunks BEGIN
    INSERT INTO chunks_fts(chunks_fts, rowid, content)
    VALUES ('delete', old.rowid, old.content);
    INSERT INTO chunks_fts(rowid, content) VALUES (new.rowid, new.content);
END;

CREATE TABLE IF NOT EXISTS meta (
    key TEXT PRIMARY KEY,
    value TEXT NOT NULL
);
`;

export class SqliteIndex {
  constructor(dbPath = ":memory:") {
    this.dbPath = String(dbPath);
    if (this.dbPath !== ":memory:") {
      mkdirSync(dirname(this.dbPath), { recursive: true });
    }
    this.db = new DatabaseSync(this.dbPath);
    this.db.exec("PRAGMA journal_mode=WAL");
    this.db.exec("PRAGMA synchronous=NORMAL");
    this.db.exec(SCHEMA);
    this._revision = this._loadRevision();
  }

  get revision() {
    return this._revision;
  }

  close() {
    this.db.close();
  }

  add(chunk) {
    this._add(chunk);
  }

  addMany(chunks) {
    let count = 0;
    for (const chunk of chunks) {
      this._add(chunk);
      count += 1;
    }
    return count;
  }

  remove(chunkId) {
    const result = this.db.prepare("DELETE FROM chunks WHERE id = ?").run(chunkId);
    if (Number(result.changes) > 0) this._bumpRevision();
  }

  clear() {
    this.db.exec("DELETE FROM chunks");
    this._bumpRevision();
  }

  get size() {
    return Number(this.db.prepare("SELECT COUNT(*) AS n FROM chunks").get().n);
  }

  get(chunkId) {
    const row = this.db.prepare("SELECT * FROM chunks WHERE id = ?").get(chunkId);
    return row ? this._rowToChunk(row) : undefined;
  }

  allChunks() {
    return this.db.prepare("SELECT * FROM chunks ORDER BY rowid").all().map((row) => this._rowToChunk(row));
  }

  chunksByKind(kind) {
    return this.db
      .prepare("SELECT * FROM chunks WHERE kind = ? ORDER BY rowid")
      .all(kind)
      .map((row) => this._rowToChunk(row));
  }

  search(query, topK = 20) {
    const ftsQuery = this._buildFtsQuery(query);
    if (ftsQuery) {
      try {
        const rows = this.db
          .prepare(
            `SELECT c.*, f.rank AS fts_rank
             FROM chunks_fts f
             JOIN chunks c ON c.rowid = f.rowid
             WHERE chunks_fts MATCH ?
             ORDER BY f.rank
             LIMIT ?`,
          )
          .all(ftsQuery, topK);
        if (rows.length > 0) return rows.map((row) => this._rowToChunk(row));
      } catch {
        // FTS syntax issue -> fall through to LIKE.
      }
    }
    const pattern = `%${query}%`;
    const rows = this.db
      .prepare("SELECT * FROM chunks WHERE content LIKE ? ESCAPE '\\' ORDER BY rowid LIMIT ?")
      .all(pattern, topK);
    return rows.map((row) => this._rowToChunk(row));
  }

  save(path = null) {
    if (path !== null && String(path) !== this.dbPath) {
      throw new Error("SqliteIndex is bound to its dbPath; use a new instance to copy.");
    }
    // DatabaseSync autocommits by default.
  }

  static load(path) {
    return new SqliteIndex(path);
  }

  _add(chunk) {
    this.db
      .prepare(
        `INSERT INTO chunks (id, source, kind, content, metadata, timestamp, priority, tokens)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?)
         ON CONFLICT(id) DO UPDATE SET
            source = excluded.source,
            kind = excluded.kind,
            content = excluded.content,
            metadata = excluded.metadata,
            timestamp = excluded.timestamp,
            priority = excluded.priority,
            tokens = excluded.tokens`,
      )
      .run(
        chunk.id,
        chunk.source,
        chunk.kind,
        chunk.content,
        JSON.stringify(chunk.metadata || {}),
        chunk.timestamp,
        chunk.priority,
        chunk.tokens,
      );
    this._bumpRevision();
  }

  _bumpRevision() {
    this._revision += 1;
    this.db
      .prepare(
        `INSERT INTO meta(key, value) VALUES('revision', ?)
         ON CONFLICT(key) DO UPDATE SET value = excluded.value`,
      )
      .run(String(this._revision));
  }

  _loadRevision() {
    const row = this.db.prepare("SELECT value FROM meta WHERE key = 'revision'").get();
    return row ? Number(row.value) : 0;
  }

  _buildFtsQuery(query) {
    const tokens = tokenize(query);
    if (tokens.length === 0) return null;
    return tokens.map((token) => `"${token}"`).join(" OR ");
  }

  _rowToChunk(row) {
    return new Chunk({
      id: row.id,
      source: row.source,
      kind: row.kind,
      content: row.content,
      metadata: JSON.parse(row.metadata || "{}"),
      timestamp: row.timestamp,
      priority: row.priority || 0,
      tokens: row.tokens || 0,
    });
  }
}

File: python/local_context_engine/sqlite_index.py
"""SQLite FTS5-backed index for high-performance local retrieval.

This is the P0 high-performance backend for Local Context Engine:

- SQLite FTS5 full-text search instead of pure-Python BM25
- disk-persistent index, WAL mode, incremental updates
- precomputed token counts stored with each chunk
- same interface as ``LocalIndex``, so the engine can swap backends

FTS5 is bundled with Python's ``sqlite3`` on most platforms (including the
WSL Python 3.12 used by this repo). If FTS5 is unavailable, the engine can
fall back to ``LocalIndex``.
"""

from __future__ import annotations

import json
import sqlite3
from pathlib import Path
from typing import Dict, Iterable, List, Optional

from .models import Chunk, SourceKind
from .text_utils import tokenize

_SCHEMA = """
CREATE TABLE IF NOT EXISTS chunks (
    rowid INTEGER PRIMARY KEY AUTOINCREMENT,
    id TEXT UNIQUE NOT NULL,
    source TEXT NOT NULL,
    kind TEXT NOT NULL,
    content TEXT NOT NULL,
    metadata TEXT NOT NULL,
    timestamp REAL,
    priority REAL,
    tokens INTEGER NOT NULL
);

CREATE VIRTUAL TABLE IF NOT EXISTS chunks_fts USING fts5(
    content,
    content='chunks',
    content_rowid='rowid',
    tokenize = 'unicode61'
);

CREATE TRIGGER IF NOT EXISTS chunks_ai AFTER INSERT ON chunks BEGIN
    INSERT INTO chunks_fts(rowid, content) VALUES (new.rowid, new.content);
END;

CREATE TRIGGER IF NOT EXISTS chunks_ad AFTER DELETE ON chunks BEGIN
    INSERT INTO chunks_fts(chunks_fts, rowid, content)
    VALUES ('delete', old.rowid, old.content);
END;

CREATE TRIGGER IF NOT EXISTS chunks_au AFTER UPDATE ON chunks BEGIN
    INSERT INTO chunks_fts(chunks_fts, rowid, content)
    VALUES ('delete', old.rowid, old.content);
    INSERT INTO chunks_fts(rowid, content) VALUES (new.rowid, new.content);
END;

CREATE TABLE IF NOT EXISTS meta (
    key TEXT PRIMARY KEY,
    value TEXT NOT NULL
);
"""


class SqliteIndex:
    """A SQLite FTS5 index exposing the same API as LocalIndex."""

    def __init__(self, db_path: str | Path = ":memory:") -> None:
        self.db_path = str(db_path)
        self._conn = sqlite3.connect(self.db_path)
        self._conn.row_factory = sqlite3.Row
        self._conn.execute("PRAGMA journal_mode=WAL")
        self._conn.execute("PRAGMA synchronous=NORMAL")
        self._conn.executescript(_SCHEMA)
        self._conn.commit()
        self._revision = self._load_revision()

    # ------------------------------------------------------------------
    # Properties / lifecycle
    # ------------------------------------------------------------------
    @property
    def revision(self) -> int:
        return self._revision

    def close(self) -> None:
        self._conn.close()

    def __enter__(self) -> "SqliteIndex":
        return self

    def __exit__(self, *exc) -> None:
        self.close()

    # ------------------------------------------------------------------
    # Mutators
    # ------------------------------------------------------------------
    def add(self, chunk: Chunk) -> None:
        self._add(chunk)
        self._conn.commit()

    def add_many(self, chunks: Iterable[Chunk]) -> int:
        count = 0
        for chunk in chunks:
            self._add(chunk)
            count += 1
        self._conn.commit()
        return count

    def remove(self, chunk_id: str) -> None:
        cur = self._conn.execute("DELETE FROM chunks WHERE id = ?", (chunk_id,))
        if cur.rowcount:
            self._bump_revision()
        self._conn.commit()

    def clear(self) -> None:
        self._conn.execute("DELETE FROM chunks")
        self._bump_revision()
        self._conn.commit()

    # ------------------------------------------------------------------
    # Accessors
    # ------------------------------------------------------------------
    def __len__(self) -> int:
        row = self._conn.execute("SELECT COUNT(*) AS n FROM chunks").fetchone()
        return int(row["n"])

    def get(self, chunk_id: str) -> Optional[Chunk]:
        row = self._conn.execute(
            "SELECT * FROM chunks WHERE id = ?", (chunk_id,)
        ).fetchone()
        return self._row_to_chunk(row) if row else None

    def all_chunks(self) -> List[Chunk]:
        rows = self._conn.execute("SELECT * FROM chunks ORDER BY rowid").fetchall()
        return [self._row_to_chunk(row) for row in rows]

    def chunks_by_kind(self, kind: SourceKind) -> List[Chunk]:
        if isinstance(kind, str):
            kind = SourceKind(kind)
        rows = self._conn.execute(
            "SELECT * FROM chunks WHERE kind = ? ORDER BY rowid", (kind.value,)
        ).fetchall()
        return [self._row_to_chunk(row) for row in rows]

    # ------------------------------------------------------------------
    # Search
    # ------------------------------------------------------------------
    def search(self, query: str, top_k: int = 20) -> List[Chunk]:
        """FTS5 search with LIKE fallback for empty/special queries."""
        fts_query = self._build_fts_query(query)
        if fts_query:
            try:
                rows = self._conn.execute(
                    """
                    SELECT c.*, f.rank AS fts_rank
                    FROM chunks_fts f
                    JOIN chunks c ON c.rowid = f.rowid
                    WHERE chunks_fts MATCH ?
                    ORDER BY f.rank
                    LIMIT ?
                    """,
                    (fts_query, top_k),
                ).fetchall()
                if rows:
                    return [self._row_to_chunk(row) for row in rows]
            except sqlite3.OperationalError:
                # FTS syntax issue -> fall through to LIKE.
                pass

        # Deterministic fallback: substring match.
        pattern = f"%{query}%"
        rows = self._conn.execute(
            """
            SELECT * FROM chunks
            WHERE content LIKE ? ESCAPE '\\'
            ORDER BY rowid
            LIMIT ?
            """,
            (pattern, top_k),
        ).fetchall()
        return [self._row_to_chunk(row) for row in rows]

    # ------------------------------------------------------------------
    # Persistence (SQLite is already persistent)
    # ------------------------------------------------------------------
    def save(self, path: Optional[str | Path] = None) -> None:
        if path is not None and str(path) != self.db_path:
            raise ValueError("SqliteIndex is bound to its db_path; use a new instance to copy.")
        self._conn.commit()

    @classmethod
    def load(cls, path: str | Path) -> "SqliteIndex":
        return cls(path)

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------
    def _add(self, chunk: Chunk) -> None:
        metadata_json = json.dumps(chunk.metadata, ensure_ascii=False)
        self._conn.execute(
            """
            INSERT INTO chunks (id, source, kind, content, metadata, timestamp, priority, tokens)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            ON CONFLICT(id) DO UPDATE SET
                source = excluded.source,
                kind = excluded.kind,
                content = excluded.content,
                metadata = excluded.metadata,
                timestamp = excluded.timestamp,
                priority = excluded.priority,
                tokens = excluded.tokens
            """,
            (
                chunk.id,
                chunk.source,
                chunk.kind.value if isinstance(chunk.kind, SourceKind) else str(chunk.kind),
                chunk.content,
                metadata_json,
                chunk.timestamp,
                chunk.priority,
                chunk.tokens,
            ),
        )
        self._bump_revision()

    def _bump_revision(self) -> None:
        self._revision += 1
        self._conn.execute(
            "INSERT INTO meta(key, value) VALUES('revision', ?) "
            "ON CONFLICT(key) DO UPDATE SET value = excluded.value",
            (str(self._revision),),
        )

    def _load_revision(self) -> int:
        row = self._conn.execute(
            "SELECT value FROM meta WHERE key = 'revision'"
        ).fetchone()
        return int(row["value"]) if row else 0

    @staticmethod
    def _build_fts_query(query: str) -> Optional[str]:
        tokens = tokenize(query)
        if not tokens:
            return None
        # OR keeps recall high; ranking still prefers documents with more terms.
        return " OR ".join(f'"{t}"' for t in tokens)

    @staticmethod
    def _row_to_chunk(row: sqlite3.Row) -> Chunk:
        return Chunk(
            id=row["id"],
            source=row["source"],
            kind=SourceKind(row["kind"]),
            content=row["content"],
            metadata=json.loads(row["metadata"] or "{}"),
            timestamp=row["timestamp"],
            priority=row["priority"] or 0.0,
            tokens=row["tokens"] or 0,
        )
